sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -178,44 +210,47 @@ private Worker(
     }
 
     private void addSpan(ReadableSpan span) {
-      if (!queue.offer(span)) {
+      if (queue.size() >= maxQueueSize) {
         droppedSpans.add(1);
+      } else {
+        queue.offer(span);
```

Comment:
I'm not comfortable with the thread-safety issues of this, it seems the queue can easily go over maxQueueSize which is an invariant we need to preserve. Since `ArrayBlockingQueue` is a queue with max capacity, let's just use it, it's defined for our job here - throughput with 0ms delay is sort of interesting theoretically but not an actual use case for this class, we can enjoy the CPU improvements when idle though thanks to the signalling which is what matters.

Reply:
Agreed queue can go over maxQueueSize constraint temporarily. How about optimistically trying to insert into queue and backing off if maxQueueSize is reached?

```
      if (size.incrementAndGet() >= maxQueueSize) {
        size.decrementAndGet();
        return false;
      }
      queue.offer(span);
      return true
``` 

concurrentQueue does have better behavior under contention with reduced exporter cpu overhead and writer thread overhead.

If MPSCQueue is the way forward as an extension, then I feel we are good with a blocking queue and batch signaling. Is it realistic to assume a batch span processor based on MPSCQueue could be supported soon?

Reply:
@sbandadd Discussed a bit with @jkwatson and let's go ahead and split this into two parts in this order

1) Replacing constant polling with signaling
2) Improve concurrency of the queuing

We don't need to bundle them into a single PR. So can we update this one to use signaling but stick with `ArrayBlockingQueue`, which you'd demonstrated does improve CPU significantly? And in a followup we can address concurrency, most likely by vendoring in `MPSCQueue` as I don't see that being a big problem.

Reply:
Sure thing. This sounds like a solid plan to me. Thanks for reaching a consensus !